src/ripple/app/misc/HashRouter.cpp

```diff
@@ -94,19 +93,27 @@ bool HashRouter::setFlags (uint256 const& key, int flags)
     return true;
 }
 
-bool HashRouter::swapSet (uint256 const& key, std::set<PeerShortID>& peers, int flag)
+bool HashRouter::prepRelay (uint256 const& key, std::set<PeerShortID>& peers)
 {
-    std::lock_guard <std::mutex> lock (mMutex);
+    std::lock_guard <std::mutex> lock (mutex_);
 
     auto& s = emplace(key).first;
 
-    if ((s.getFlags () & flag) == flag)
+    if (!s.prepRelay(suppressionMap_.clock().now(), holdTime_))
         return false;
 
-    s.swapSet (peers);
-    s.setFlags (flag);
+    s.swap (peers);
 
     return true;
 }
 
+bool HashRouter::prepRelay (uint256 const& key)
+{
+    std::lock_guard <std::mutex> lock (mutex_);
+
+    auto& s = emplace(key).first;
+
+    return s.prepRelay(suppressionMap_.clock().now(), holdTime_);
+}
```

Comment:
The answer to your question is this change, which is one of the main reasons for the `HashRouter` tracking relay time separately in the first place.
https://github.com/ripple/rippled/pull/1749/files#diff-c71c719ae07d12ad17a81571fcea7f26L1306

Before, all it did was set the flag and relayed to all peers if the flag wasn't set before. _Buuut_ I think that's because it assumed knowledge of the behavior of `HashRouter` - the only way that `setFlags` would return true is if it created a new entry, which would thus have an empty peer list.

...So thanks to your and @seelabs's questions, I'm rethinking that `prepRelay` should only have one override so the stored peer list gets cleared regardless, then leave it up to the client whether to ignore the peer list or not. Maybe it should return a `pair` instead of taking the reference parameter, too.